### PR TITLE
adding datasource and resource for ds auth provider certificate

### PR DESCRIPTION
**Pull request by sapana05**
ds auth provider certificate resource and datasource

##### ISSUE TYPE
- Feature Pull Request

#### Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

### CODE REVIEW

**Comment by terraform-providers-svc** [comment]
Can one of the admins verify this patch?

File: gofish/dell/directoryServiceAuthProviderCertificate.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	Organization       string `json:"@odata.context"`
+	City               string `json:"@odata.context"`
+	Country            string `json:"@odata.context"`
+	Email              string `json:"@odata.context"`
+	OrganizationalUnit string `json:"@odata.context"`
+}
+
+type CertificateCollection struct {
+	OdataContext string `json:"@odata.context"`
+	OdataID      string `json:"@odata.id"`
```
Comment (taohe1012) [inline gofish/dell/directoryServiceAuthProviderCertificate.go:48]: hi @sapana05 , if those params are useless, please remove lines

File: redfish/provider/data_source_redfish_directory_service_auth_provider_certificate.go [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+		"odata_id": schema.StringAttribute{
+			MarkdownDescription: "OData ID for the Certificate",
+			Description:         "OData ID for the Certificate",
+			Computed:            true,
+		},
+		/*"id": schema.StringAttribute{
+			MarkdownDescription: "ID of the Certificate",
+			Description:         "ID of the Certificate",
+			Computed:            true,
+		},*/
```
Comment (taohe1012) [inline redfish/provider/data_source_redfish_directory_service_auth_provider_certificate.go:151]: same

File: redfish/provider/data_source_redfish_directory_service_auth_provider_certificate.go [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+	d.ID = types.StringValue(fmt.Sprintf("%d", time.Now().Unix()))
+
+	dellCertificate, certErr := dell.DirectoryServiceAuthProvider(accountService)
+
+	if certErr != nil {
+		diags.AddError("Unable to fetch Certificate URI", "Unable to fetch Certificate URI")
+	}
+
+	var certificateURI string
+	//diags.AddError("CertificateProviderType", d.CertificateFilter.CertificateProviderType.String())
```
Comment (taohe1012) [inline redfish/provider/data_source_redfish_directory_service_auth_provider_certificate.go:297]: same

File: redfish/provider/data_source_redfish_directory_service_auth_provider_certificate.go [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+
+	if !d.CertificateFilter.CertificateId.IsNull() && !d.CertificateFilter.CertificateId.IsUnknown() {
+		certificateDetailsURI = certificateURI + "/" + d.CertificateFilter.CertificateId.ValueString()
+	}
+
+	//diags.AddError("certificateDetailsURI", certificateDetailsURI)
+	certResponse, err := g.service.GetClient().Get(certificateDetailsURI)
+
+	if err != nil {
+		//diags.AddError("certificateDetailsURI ", " : "+certificateDetailsURI)
```
Comment (taohe1012) [inline redfish/provider/data_source_redfish_directory_service_auth_provider_certificate.go:359]: same

File: redfish/provider/data_source_redfish_directory_service_auth_provider_certificate.go [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+			return d, diags
+		}
+
+	}
+
+	if !d.CertificateFilter.CertificateId.IsNull() && !d.CertificateFilter.CertificateId.IsUnknown() {
+		certificateDetailsURI = certificateURI + "/" + d.CertificateFilter.CertificateId.ValueString()
+	}
+
+	//diags.AddError("certificateDetailsURI", certificateDetailsURI)
```
Comment (taohe1012) [inline redfish/provider/data_source_redfish_directory_service_auth_provider_certificate.go:355]: same

File: redfish/provider/data_source_redfish_directory_service_auth_provider_certificate.go [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+	if err != nil {
+		return d, diags
+	}
+
+	err = json.Unmarshal(certBody, &certificate)
+	if err != nil {
+		diags.AddError("Error parsing Certificate", err.Error())
+		return d, diags
+	}
+	/*if 1 == 1 {
```
Comment (taohe1012) [inline redfish/provider/data_source_redfish_directory_service_auth_provider_certificate.go:378]: same

File: redfish/provider/data_source_redfish_directory_service_auth_provider_certificate.go [hunk 7]
```diff
@@ -0,0 +1,480 @@
...
+		prefixValue := "SecurityCertificate." + middleindex
+		//diags.AddError("prefixValue", prefixValue)
+		for k, v := range idracAttributesState.Attributes.Elements() {
+			if strings.HasPrefix(k, prefixValue) {
+				attributesToReturn[k] = v
+			}
+		}
+	}
+	securityValue := types.MapValueMust(types.StringType, attributesToReturn)
+	/*if d.SecurityCertificate.IsNull() || d.SecurityCertificate.IsUnknown() {
```
Comment (taohe1012) [inline redfish/provider/data_source_redfish_directory_service_auth_provider_certificate.go:426]: same

File: redfish/provider/resource_redfish_directory_service_auth_provider_certificate.go [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+
+	api, err := NewConfig(r.p, &plan.RedfishServer)
+	if err != nil {
+		resp.Diagnostics.AddError("service error", err.Error())
+		return
+	}
+	service := api.Service
+	defer api.Logout()
+
+	if plan.CertificateType.IsNull() || plan.CertificateType.IsUnknown() {
```
Comment (taohe1012) [inline redfish/provider/resource_redfish_directory_service_auth_provider_certificate.go:135]: we can use validator to do this check

File: redfish/provider/resource_redfish_directory_service_auth_provider_certificate.go [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+		resp.Diagnostics.AddError("required config missing", "please provide valid config for required param")
+	}
+
+	if plan.CertificateType.ValueString() != pem && plan.CertificateType.ValueString() != rsaCACert {
```
Comment (taohe1012) [inline redfish/provider/resource_redfish_directory_service_auth_provider_certificate.go:139]: we can use validator to do this check

**Comment by sapana05** [comment]
image
image
image

File: redfish/provider/resource_redfish_directory_service_auth_provider_certificate.go [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+	}
+
+	api, err := NewConfig(r.p, &plan.RedfishServer)
+	if err != nil {
+		resp.Diagnostics.AddError("service error", err.Error())
+		return
+	}
+	service := api.Service
+	defer api.Logout()
+	if plan.CertificateType.ValueString() != pem && plan.CertificateType.ValueString() != rsaCACert {
```
Comment (taohe1012) [inline redfish/provider/resource_redfish_directory_service_auth_provider_certificate.go:135]: the CertificateType is already in the validator, could we remove this check?

File: redfish/provider/resource_redfish_directory_service_auth_provider_certificate.go [hunk 11]
```diff
@@ -0,0 +1,398 @@
...
+	}
+	api, err := NewConfig(r.p, &plan.RedfishServer)
+	if err != nil {
+		resp.Diagnostics.AddError("service error", err.Error())
+		return
+	}
+	service := api.Service
+	defer api.Logout()
+
+	if plan.CertificateType.ValueString() != pem && plan.CertificateType.ValueString() != rsaCACert {
```
Comment (taohe1012) [inline redfish/provider/resource_redfish_directory_service_auth_provider_certificate.go:208]: same

**Review by taohe1012 — DISMISSED** [review]

**Review by MangirishK — APPROVED** [review]

**Review by taohe1012 — APPROVED** [review]

**Pull request merged**